src/theory/shared_solver.h

```diff
@@ -0,0 +1,135 @@
+/*********************                                                        */
+/*! \file shared_solver.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Base class for shared solver
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef CVC4__THEORY__SHARED_SOLVER__H
+#define CVC4__THEORY__SHARED_SOLVER__H
+
+#include "expr/node.h"
+#include "theory/ee_setup_info.h"
+#include "theory/logic_info.h"
+#include "theory/shared_terms_database.h"
+#include "theory/term_registration_visitor.h"
+#include "theory/valuation.h"
+
+namespace CVC4 {
+
+class TheoryEngine;
+
+namespace theory {
+
+/**
+ * A base class for shared solver. The shared solver is the component of theory
+ * engine that behaves like a theory solver, and whose purpose is to ensure the
+ * main theory combination method can be performed in CombinationEngine.
+ * Its role is to:
+ * (1) Notify the individual theories of shared terms via addSharedTerms,
+ * (2) Be the official interface for equality statuses,
+ * (3) Propagate equalities to TheoryEngine when necessary and explain them.
+ */
+class SharedSolver
+{
+ public:
+  SharedSolver(TheoryEngine& te);
+  virtual ~SharedSolver() {}
+  //------------------------------------- initialization
+  /**
+   * Returns true if we need an equality engine, this has the same contract
+   * as Theory::needsEqualityEngine.
+   */
+  virtual bool needsEqualityEngine(theory::EeSetupInfo& esi);
+  /**
+   * Set the equality engine. This should be called by equality engine manager
+   * during EqEngineManager::initializeTheories.
+   */
+  virtual void setEqualityEngine(eq::EqualityEngine* ee) = 0;
+  //------------------------------------- end initialization
+  /**
+   * Called when the given term t is pre-registered in TheoryEngine.
+   *
+   * This adds t as an equality to propagate in the shared terms database
+   * if it is an equality, or adds its shared terms if it involves multiple
+   * theories.
+   *
+   * @param t The term to preregister
+   * @param multipleTheories Whether multiple theories are present in t.
+   */
+  void preRegisterShared(TNode t, bool multipleTheories);
+  /**
+   * Pre-notify assertion fact with the given atom. This is called when any
+   * fact is asserted in TheoryEngine, just before it is dispatched to the
+   * appropriate theory.
+   *
+   * This calls Theory::notifySharedTerm for the shared terms of the atom.
+   */
+  void preNotifySharedFact(TNode atom);
+  /**
+   * Get the equality status of a and b, which first asks if the shared
+   * terms database as an equality status, and otherwise asks the appropriate
+   * Theory.
```

Comment:
The base class doesn't actually do that, only the `SharedSolverDistributed`. Maybe just say "Get the equality status of a and b"?

Reply:
Ah, right, I'll update.